File: 1876-substrings-of-size-three-with-distinct-characters/1876-substrings-of-size-three-with-distinct-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var countGoodSubstrings = function(s) {
    const freq = new Array(26).fill(0);
    let l = 0;
    let r = 0;
    let count = 0;
    const ch = s.split('');
    while (r < ch.length) {
        while (l <= r && r - l + 1 > 3) {
            freq[ch[l].charCodeAt(0) - 'a'.charCodeAt(0)]--;
            l++;
        }
        freq[ch[r].charCodeAt(0) - 'a'.charCodeAt(0)]++;
        if (r - l + 1 == 3 
                && freq[ch[l].charCodeAt(0) - 'a'.charCodeAt(0)] == 1
                && freq[ch[r].charCodeAt(0) - 'a'.charCodeAt(0)] == 1
            )
            count++;
        r++;
    }
    return count;
};

File: 1876-substrings-of-size-three-with-distinct-characters/1876-substrings-of-size-three-with-distinct-characters.java
class Solution {
    public int countGoodSubstrings(String s) {
        int[] freq = new int[26];
        
        int l = 0, r = l;
        int count = 0;

        char[] ch = s.toCharArray();

        while (r < ch.length) {
            while (l <= r && r - l + 1 > 3) {
                freq[ch[l] - 'a']--;
                l++;
            }

            freq[ch[r] - 'a']++;

            if (r - l + 1 == 3 
                    && freq[ch[l] - 'a'] == 1 
                    && freq[ch[r] - 'a'] == 1)
                count++;

            r++;
        }

        return count;
    }
}